File: Practica 5/EcuacionPOO.java
import java.util.Scanner;

abstract class EcuacionCuadratica {
    double a, b, c, discriminante;

    EcuacionCuadratica(double a, double b, double c) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.discriminante = b * b - 4 * a * c;
    }

    abstract String calcular();
}

class DosRaices extends EcuacionCuadratica {
    DosRaices(double a, double b, double c) {
        super(a, b, c);
    }

    @Override
    String calcular() {
        double r1 = (-b + Math.sqrt(discriminante)) / (2 * a);
        double r2 = (-b - Math.sqrt(discriminante)) / (2 * a);
        return String.format("La ecuación tiene dos raíces %.5f y %.5f", r1, r2);
    }
}

class UnaRaiz extends EcuacionCuadratica {
    UnaRaiz(double a, double b, double c) {
        super(a, b, c);
    }

    @Override
    String calcular() {
        double r = -b / (2 * a);
        return String.format("La ecuación tiene una raíz %.5f", r);
    }
}

class SinRaices extends EcuacionCuadratica {
    SinRaices(double a, double b, double c) {
        super(a, b, c);
    }

    @Override
    String calcular() {
        return "La ecuación no tiene raíces reales";
    }
}

public class EcuacionPOO {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.print("Ingrese a, b, c (o escriba 'salir' para terminar): ");
            String entrada = scanner.nextLine();

            if (entrada.equalsIgnoreCase("salir")) break;

            try {
                String[] valores = entrada.split(" ");
                double a = Double.parseDouble(valores[0]);
                double b = Double.parseDouble(valores[1]);
                double c = Double.parseDouble(valores[2]);

                EcuacionCuadratica ecuacion = (a == 0) ? new SinRaices(a, b, c) :
                        (b * b - 4 * a * c > 0) ? new DosRaices(a, b, c) :
                        (b * b - 4 * a * c == 0) ? new UnaRaiz(a, b, c) : new SinRaices(a, b, c);

                System.out.println(ecuacion.calcular());
            } catch (Exception e) {
                System.out.println("Entrada inválida.");
            }
        }

        scanner.close();
    }
}

File: Practica 5/ecuacionPOO.py
import math

class EcuacionCuadratica:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c
        self.discriminante = self.getDiscriminante()

    def getDiscriminante(self):
        return self.b**2 - 4 * self.a * self.c

    def resolver(self):
        if self.discriminante > 0:
            return DosRaices(self.a, self.b, self.discriminante).calcular()
        elif self.discriminante == 0:
            return UnaRaiz(self.a, self.b).calcular()
        else:
            return SinRaices().calcular()

class DosRaices(EcuacionCuadratica):
    def calcular(self):
        r1 = (-self.b + math.sqrt(self.discriminante)) / (2 * self.a)
        r2 = (-self.b - math.sqrt(self.discriminante)) / (2 * self.a)
        return f"La ecuación tiene dos raíces {r1:.5f} y {r2:.5f}"

class UnaRaiz(EcuacionCuadratica):
    def calcular(self):
        r = -self.b / (2 * self.a)
        return f"La ecuación tiene una raíz {r:.5f}"

class SinRaices(EcuacionCuadratica):
    def calcular(self):
        return "La ecuación no tiene raíces reales"

while True:
    entrada = input("Ingrese a, b, c (o escriba 'salir' para terminar): ")
    if entrada.lower() == "salir":
        print("Programa terminado.")
        break
    try:
        a, b, c = map(float, entrada.split())
        ecuacion = EcuacionCuadratica(a, b, c)
        print(ecuacion.resolver())
    except ValueError:
        print("Entrada inválida. Asegúrese de ingresar tres números separados por espacios.")
